### DISCUSSION TITLE
AWS systems_manager ParameterNotFound error catching

### DISCUSSION THREAD

**Discussion by nsepetys** [apache/airflow#14707] [Q&A] [answered]
Hi,

I am wondering if anyone has information as to why [AWS secrets handling](https://github.com/apache/airflow/blob/master/airflow/providers/amazon/aws/secrets/systems_manager.py#L142-L148) for the `ParameterNotFound` client exception does not get raised? This error should be something Airflow administrators have taken care of to avoid in the first place but accidents happen and these parameters can end up removed or the names changes. The result is that tasks don't fail and the error slips through the cracks causing a downstream trickle of unexpected situations. This is not expected primarily because when choosing the default Airflow secrets back-end we get a different experience when the key is missing- the issue is raised.

- [Variables.get](https://github.com/apache/airflow/blob/master/airflow/models/variable.py#L132) for default raises KeyError
  - To my best understanding AWS secrets back-end passes through this method but outcome is different depending on whether or not AWS is used
- [Connections.get_connection_from_secrets](https://github.com/apache/airflow/blob/a9ac2b040b64de1aa5d9c2b9def33334e36a8d22/airflow/models/connection.py#L354) raises AirflowNotFoundException
  - Again, according to my best understanding, the secrets back-end (regardless of whether it's AWS or default) comes through here but the raise never happens here because of the `ParameterNotFound` handling earlier on.
  
@kaxil - You appear to be the authority on what is there (basing this on one of the [latest PRs](https://github.com/apache/airflow/pull/7745) in AWS systems manager sphere). Thank you for your work on all this. It has made our org happy to have a better method for handling secrets. I am curious though if we had any reason why we have the approach we currently have for this particular error catch. I am somewhat new to Airflow internals and also somewhat new to Python, so I'm hoping, if anything, you can help educate me.

**Comment by kaxil** [comment, marked as answer]
Hi @nsepetys,

Glad that you found the secrets backend useful.

The main reason we don't raise `ParameterNotFound` for AWS secrets backend ([`SystemsManagerParameterStoreBackend`](https://github.com/apache/airflow/blob/ffe3bd29574d62a0a692cd8f63995856bbff8c0b/airflow/providers/amazon/aws/secrets/systems_manager.py#L142) and [ `SecretsManagerBackend`](https://github.com/apache/airflow/blob/ffe3bd29574d62a0a692cd8f63995856bbff8c0b/airflow/providers/amazon/aws/secrets/secrets_manager.py#L156)) is so that we can fallback to getting variables or Connections via Environment Variables and then Metadata Database.

For example, for Airflow Variables:
![image](https://user-images.githubusercontent.com/8811558/110711160-f65a8f00-81f6-11eb-93fc-bea1e210a286.png)

We do a similar thing for Hashicorp Vault and GoogleCloud Secrets Backend but as they don't raise an error, we just return `None` (e.g https://github.com/apache/airflow/blob/ffe3bd29574d62a0a692cd8f63995856bbff8c0b/airflow/providers/google/cloud/secrets/secret_manager.py#L148-L158)

We loop across all the 3 backends (Secrets Backend in the airflow.cfg, Environment Variables and then MetastoreBackend)
https://github.com/apache/airflow/blob/a9ac2b040b64de1aa5d9c2b9def33334e36a8d22/airflow/models/connection.py#L343-L354

https://github.com/apache/airflow/blob/ffe3bd29574d62a0a692cd8f63995856bbff8c0b/airflow/configuration.py#L1017-L1033

We only raise an error when Connection or Variable is not found in all the 3 backends

**Reply by nsepetys** [reply]
Thanks for taking the time to explain @kaxil . That provided some good details. I think this new understanding has given me a theory as to what is going on with our instance of Airflow. You take care!
